### PR TITLE
Use SCM_USERNAME and SCM_ACCESS_TOKEN to checkout

### PR DESCRIPTION
**Pull request by d2lam**
Use SCM_USERNAME and SCM_ACCESS_TOKEN to checkout
These can be set as secrets. If they are not set, then just checkout normally

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -161,14 +161,21 @@ class GithubScm extends Scm {
      * @return {Promise}
      */
     _getCheckoutCommand(config) {
-        const checkoutUrl = `https://${config.host}/${config.org}/${config.repo}`;
+        const checkoutUrl = `${config.host}/${config.org}/${config.repo}`;
         const checkoutRef = config.prRef ? config.branch : config.sha; // if PR, use pipeline branch
         const command = [];
+        const checkoutCommand = [
```
Comment (minz1027) [inline index.js]: string?

**Comment by tkyi** [comment]
Currently we are echoing out all the commands when we run steps (outside of sd-setup). I don't think we would want to echo out the SCM_USERNAME and SCM_ACCESS_TOKEN in the build logs.

**Comment by minz1027** [comment]
To add to @tkyi 's comment, the executor run is printing out the command. :(
https://github.com/screwdriver-cd/launcher/blob/master/executor/executor.go#L38

**Comment by d2lam** [comment]
It won't evaluate the ENV.

**Comment by tkyi** [comment]
Oh actually, looks like it does not evaluate the environment variables when it echoes, so it should be okay 👍

File: index.js (hunk 1 above)
Comment (d2lam) [inline index.js]: Fixed

**Comment by FenrirUnbound** [comment]
This will temporarily resolve our issue. What does the long-term solution look like?

File: index.js [hunk 2]
```diff
@@ -161,14 +161,20 @@ class GithubScm extends Scm {
...
     _getCheckoutCommand(config) {
-        const checkoutUrl = `https://${config.host}/${config.org}/${config.repo}`;
+        const checkoutUrl = `${config.host}/${config.org}/${config.repo}`;
         const checkoutRef = config.prRef ? config.branch : config.sha; // if PR, use pipeline branch
         const command = [];
+        const checkoutCommand =
+            'if [ -z $SCM_USERNAME ] || [ -z $SCM_ACCESS_TOKEN ]; then ' +
+            `git clone --quiet --progress --branch ${config.branch} ` +
+            `https://${checkoutUrl} $SD_SOURCE_DIR; ` +
+            'else ' +
```
Comment (stjohnjohnson) [inline index.js]: I'm not a fan of having this complexity in here.  We really only need one clone command, but something that creates the `$SCM_USERNAME:$SCM_ACCESS_TOKEN@` part and injects it if necessary.

Comment (stjohnjohnson) [inline index.js]: Something like (probably doesn't work).

```
SCM_URL=${checkoutUrl}
if [ ! -z $SCM_USERNAME ] && [ ! -z $SCM_ACCESS_TOKEN ]; then   SCM_URL="$SCM_USERNAME:$SCM_ACCESS_TOKEN@$SCM_URL";
fi
git clone --quiet --progress --branch ${config.branch} https://$SCM_URL $SD_SOURCE_DIR
```

Comment (d2lam) [inline index.js]: Okay fixed.

File: index.js [hunk 3]
```diff
@@ -161,14 +161,19 @@ class GithubScm extends Scm {
...
     _getCheckoutCommand(config) {
-        const checkoutUrl = `https://${config.host}/${config.org}/${config.repo}`;
+        const checkoutUrl = `${config.host}/${config.org}/${config.repo}`;
         const checkoutRef = config.prRef ? config.branch : config.sha; // if PR, use pipeline branch
         const command = [];
+        const checkoutCommand =
+            `export SCM_URL=${checkoutUrl}` +
+            '[[ -n "$SCM_USERNAME" && -n "$SCM_ACCESS_TOKEN" ]] ' +
+            '&& SCM_URL="$SCM_USERNAME:$SCM_ACCESS_TOKEN@$SCM_URL"' +
+            `git clone --quiet --progress --branch ${config.branch} ` +
```
Comment (stjohnjohnson) [inline index.js]: You are going to need newlines, no?

**Comment by d2lam** [comment]
@stjohnjohnson  should work now

**Comment by d2lam** [comment]
I echo out the $SCM_URL to see if it's cloning the correct value:

screen shot 2016-11-22 at 11 29 53 am

**Pull request merged**